Refactor/integrate gn proxy auth mechanism

BonfaceKilz opened the pull request:
#### Description
<!--Brief description of the PR. What does this PR do? -->
This PR:

- Deletes some unused logger imports
- Moves markdown_routes to api/markdown. This is more semantic.
- Adds a default setting to our configs for REDIS
- Deletes dead code in redis_utils
- Makes GN2_PROXY a configurable option
- Removes old routes for managing resources
- Adds an OrderedEnum class that's used to represent roles.
- Add extra decorators:
  - `@login_required`
  -  `@edit_access_required`
- Adds the following endpoints:
  - GET /resource-management/resources/<resource_id>
  - POST /resource-management/resources/<resource-id>/make-public
  - POST /resources/<resource_id>/change-owner
 
#### How should this be tested?
<!-- What should you do to test this PR? Is there any manual quality
assurance checks that should be done. What are the expectations -->
At the moment, this can't be tested. The user's hash and resources column need to be looked at again before we use this!

#### Any background context you want to provide?
<!-- Anything the reviewer should be aware of ahead of testing -->
See: https://github.com/genenetwork/gn-gemtext-threads/blob/main/issues/authorisation.gmi

#### What are the relevant pivotal tracker stories?
<!-- Does this PR track anything anywhere? -->
https://github.com/genenetwork/gn-gemtext-threads/blob/main/issues/authorisation.gmi

#### Screenshots (if appropriate)

![image](https://user-images.githubusercontent.com/11820306/137921094-5e8ac9b6-be0f-4ac6-a79f-5b508b00d252.png)

![image](https://user-images.githubusercontent.com/11820306/138674635-e5db0b72-1e5d-4a8a-9518-66e3b42c8579.png)


#### Questions
<!-- Are there any questions for the reviewer -->

N/A

BonfaceKilz committed:
api: correlation: Delete unused logger

BonfaceKilz committed:
api: correlation: Delete unused imports

BonfaceKilz committed:
Move markdown_routes to api/markdown

All new API definitions should be migrated to "wqflask/ api"

BonfaceKilz committed:
Add default setting for REDIS_URL

BonfaceKilz committed:
Mark `get_resource_info` as deprecated

BonfaceKilz committed:
Add test-cases for `get_user_membership`

BonfaceKilz committed:
resource_manager: Delete every route & helper method in this module

This module is coupled to `utility.redis_tools` which needs to be
deprecated. Also, the routes are created by importing `app` instead of
creating a new blueprint.

BonfaceKilz committed:
resource_manager: Add `get_user_membership` method

BonfaceKilz committed:
resource_manager: Add OrderedEnum

Borrowed from:
<https://github.com/woodruffw/ordered_enum/tree/master/src/ordered_enum>

BonfaceKilz committed:
Add immutable data structures for admin and data access roles

BonfaceKilz committed:
Get a users access mask

If a user has several access roles, select the highest role.

zsloan commented:
One minor thing is that I think the default mask would actually be something like view/view/not-admin (for data/metadata/admin respectively). Most of the time data will be public, so the default is for users to be able to view it (but obviously not edit it).

In line 105 how exactly is the "max(v)" working? Is that making use of the `__lt__` in OrderedEnum to compare indices of each access level?

It's interesting and helpful seeing the way you do something like this. It's much better than before, but in a way that is easy to understand and often makes me think "why didn't I do things like that?"

BonfaceKilz commented:
zsloan ***@***.***> anaandika:

> One minor thing is that I think the default mask
> would actually be something like
> view/view/not-admin (for data/metadata/admin
> respectively).

Nice observation there:

```
    access_role = {
        "data": [DataRole.NO_ACCESS],
        "metadata": [DataRole.NO_ACCESS],
        "admin": [AdminRole.NOT_ADMIN],
    }
```

I use the lowest(!) access-role in the data there
is. That's because, as is, every resource has a
default mask which sets the access-role to "VIEW"
anyways. In other words, I'm assuming that in the
case of no "default-mask", the data should not be
accessible. Which is a good IMHO assumption to
make.

> Most of the time data will be
> public, so the default is for users to be able to
> view it (but obviously not edit it).

See above.

> In line 105 how exactly is the "max(v)" working?
> Is that making use of the __lt__ in OrderedEnum to
> compare indices of each access level?

Ah that. I prefer immutable types, hence I use the
`Enum` class. However, by default, the `Enum` type
does not have any way of comparing it's elements
unless they are integers. So for that I use the
OrderedEnum. As an example, with `OrderedEnum`
with this example:

```
@unique
class DataRole(OrderedEnum):
    NO_ACCESS = "no-access"
    VIEW = "view"
    EDIT = "edit"
```

`max(Datarole.NO_ACCESS, Datarole.VIEW,
Datarole.EDIT)`, will yield "Datorole.NO_ACCESS"
with it's string repr being "no-access".

> It's interesting and helpful seeing the way you do
> something like this. It's much better than before,
> but in a way that is easy to understand and often
> makes me think "why didn't I do things like that?"

You get better with time ;) ;)

-- 
Bonface M. K.
D4F09EB110177E03C28E2FE1F5BBAE1E0392253F (hkp://keys.gnupg.net)
Free Software Activist
Humble GNU Emacs User | Bearer of scheme-y parens
Curator: <https://upbookclub.com> | Twitter: @BonfaceKilz

BonfaceKilz committed:
Replace "resource_id" with "resource_info" dict

This avoids calling Redis twice when fetching metadata about the
resource.

BonfaceKilz committed:
Remove "utility.hmac.hmac_creation" which causes circular imports

Hacky but re-implement `hmac_creation` as `create_hmac`

BonfaceKilz committed:
decorators: Add `@login_required` decorator

BonfaceKilz committed:
utility.hmac: Label "hmac_creation" as deprecated

This function is coupled to "wqflask.app", therefore requiring it's
import at the module level. This may lead circular importation issues
when working with blueprints.

BonfaceKilz committed:
manage_resource.html: Remove "flash"

BonfaceKilz committed:
manage_resource.html: If owner owns a resource display their name

BonfaceKilz committed:
manage_resource.html: Display group name next to it's unique id

BonfaceKilz committed:
Remove dead line

BonfaceKilz committed:
wqflask: resource_manager: Add `add_extra_resource_metadata` method

BonfaceKilz committed:
Add "GET /resource-management/resources/<resource_id>"

BonfaceKilz committed:
authentication_tools: Mark `check_owner_or_admin` as deprecated

Use the new auth proxy tools instead.

BonfaceKilz committed:
manage_resource.html: Remove unused stylesheets

BonfaceKilz committed:
manage_resource: Add missing bootstrap "table" class

BonfaceKilz committed:
manage_resource: Remove column "Admin" from group permissions table

A group does not have "admin" privileges. Instead, a user can have
admin privileges over a particular group.

BonfaceKilz committed:
manage_resources.html: Clean up html

BonfaceKilz committed:
wqflask: decorators: Fix type when calling hexdigest()

BonfaceKilz committed:
scripts: group.py: Add shebang

BonfaceKilz committed:
scripts: group.py: Add extra optional arg to specify uid

BonfaceKilz committed:
scripts: resource.py: Add shebang

BonfaceKilz committed:
Move access_roles enums to it's own module

BonfaceKilz committed:
wqflask: decorators: Make `edit_access_required` more generic

This now supports passing the `resource_id` directly.

BonfaceKilz committed:
manage_resource.html: Toggle the "is open to public" radio buttons

BonfaceKilz committed:
Rename manage_resource -> view_resource

BonfaceKilz committed:
resource_manager: Use := to store embellished resource dict

BonfaceKilz committed:
wqflask: resource_manager: Add "resource_id" to resource dict

BonfaceKilz committed:
Add `POST resource-management/resources/<resource-id>/make-public`

This endpoint either makes resources public or non-public by tweaking
the access-masks.

BonfaceKilz committed:
scripts: resource: Add value for "admin" when updating resources

BonfaceKilz committed:
Add "GN2_PROXY" as a configurable option

BonfaceKilz committed:
Add `edit_admins_required` decorator

BonfaceKilz committed:
Use gn-proxy to fetch access-roles instead of looking at redis

BonfaceKilz committed:
Update failing tests when testing `get_user_access_roles`

BonfaceKilz committed:
decorators: Use the correct logic to check for edit_admins_access

BonfaceKilz committed:
manage_resource.html: Update the URL for updating the owner

BonfaceKilz committed:
utility: redis_tools: Remove dead functions

BonfaceKilz committed:
Add "GET /resources/<resource_id>/change-owner" endpoint

BonfaceKilz committed:
Add "POST /resource-management/<resource_id>/users/search"

BonfaceKilz committed:
change_resource_owner.html: Update endpoint for user-search

This link has the {{ resource_id }} set because it's a required
argument for the "@edit_admins_access_required" decorator

BonfaceKilz committed:
change_resource_owner.html: Replace "var" with "let"

"let" is lexically scoped.

BonfaceKilz committed:
change_resource_owner.html: Replace "the_html" with searchResultHtml

BonfaceKilz committed:
Add "POST /resources/<resource_id>/change-owner" endpoint

BonfaceKilz committed:
manage_resource.html: Replace button with link

BonfaceKilz commented:
@zsloan let's avoid using utility.tools to define configuration options; and instead use default_settings.py because of the following circular imports issues:

```
Traceback (most recent call last):
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/gnu/store/286imikhis0dvb5aqcy87q6x8542xpw9-gunicorn-20.1.0/lib/python3.8/site-packages/gunicorn/util.py", line 359, in import_app
    mod = importlib.import_module(module)
  File "/gnu/store/sd2ic7bpv8fx3imy1j8xcjclx71sv6q0-python-3.8.2/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/bonface/projects/genenetwork2/wqflask/wsgi.py", line 1, in <module>
    from run_gunicorn import app as application  # expect application as a name
  File "/home/bonface/projects/genenetwork2/wqflask/run_gunicorn.py", line 12, in <module>
    from wqflask import app
  File "/home/bonface/projects/genenetwork2/wqflask/wqflask/__init__.py", line 12, in <module>
    from wqflask.resource_manager import resource_management
  File "/home/bonface/projects/genenetwork2/wqflask/wqflask/resource_manager.py", line 19, in <module>
    from wqflask.decorators import edit_access_required
  File "/home/bonface/projects/genenetwork2/wqflask/wqflask/decorators.py", line 12, in <module>
    from utility.tools import GN_PROXY_URL
  File "/home/bonface/projects/genenetwork2/wqflask/utility/tools.py", line 8, in <module>
    from wqflask import app
ImportError: cannot import name 'app' from partially initialized module 'wqflask' (most likely due to a circular import) (/home/bonface/projects/genenetwork2/wqflask/wqflask/__init__.py)
```

ATM, with this PR, there are 2 different places where the auth proxy are defined...
